File: config/src/main/java/com/typesafe/config/impl/ConfigDelayedMerge.java
/**
 *   Copyright (C) 2011-2012 Typesafe Inc. <http://typesafe.com>
 */
package com.typesafe.config.impl;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

import com.typesafe.config.ConfigException;
import com.typesafe.config.ConfigOrigin;
import com.typesafe.config.ConfigRenderOptions;
import com.typesafe.config.ConfigValueType;

/**
 * The issue here is that we want to first merge our stack of config files, and
 * then we want to evaluate substitutions. But if two substitutions both expand
 * to an object, we might need to merge those two objects. Thus, we can't ever
 * "override" a substitution when we do a merge; instead we have to save the
 * stack of values that should be merged, and resolve the merge when we evaluate
 * substitutions.
 */
final class ConfigDelayedMerge extends AbstractConfigValue implements Unmergeable,
        ReplaceableMergeStack {

    // earlier items in the stack win
    final private List<AbstractConfigValue> stack;

    ConfigDelayedMerge(ConfigOrigin origin, List<AbstractConfigValue> stack) {
        super(origin);
        this.stack = stack;
        if (stack.isEmpty())
            throw new ConfigException.BugOrBroken(
                    "creating empty delayed merge value");

        for (AbstractConfigValue v : stack) {
            if (v instanceof ConfigDelayedMerge || v instanceof ConfigDelayedMergeObject)
                throw new ConfigException.BugOrBroken(
                        "placed nested DelayedMerge in a ConfigDelayedMerge, should have consolidated stack");
        }
    }

    @Override
    public ConfigValueType valueType() {
        throw new ConfigException.NotResolved(
                "called valueType() on value with unresolved substitutions, need to Config#resolve() first, see API docs");
    }

    @Override
    public Object unwrapped() {
        throw new ConfigException.NotResolved(
                "called unwrapped() on value with unresolved substitutions, need to Config#resolve() first, see API docs");
    }

    @Override
    ResolveResult<? extends AbstractConfigValue> resolveSubstitutions(ResolveContext context, ResolveSource source)
            throws NotPossibleToResolve {
        return resolveSubstitutions(this, stack, context, source);
    }

    // static method also used by ConfigDelayedMergeObject
    static ResolveResult<? extends AbstractConfigValue> resolveSubstitutions(ReplaceableMergeStack replaceable,
            List<AbstractConfigValue> stack,
            ResolveContext context, ResolveSource source) throws NotPossibleToResolve {
        if (ConfigImpl.traceSubstitutionsEnabled()) {
            ConfigImpl.trace(context.depth(), "delayed merge stack has " + stack.size() + " items:");
            int count = 0;
            for (AbstractConfigValue v : stack) {
                ConfigImpl.trace(context.depth() + 1, count + ": " + v);
                count += 1;
            }
        }

        // to resolve substitutions, we need to recursively resolve
        // the stack of stuff to merge, and merge the stack so
        // we won't be a delayed merge anymore. If restrictToChildOrNull
        // is non-null, or resolve options allow partial resolves,
        // we may remain a delayed merge though.

        ResolveContext newContext = context;
        int count = 0;
        AbstractConfigValue merged = null;
        for (AbstractConfigValue end : stack) {
            // Per the HOCON spec, a substitution hidden by a value that
            // cannot be merged with it is never evaluated. If merged already
            // ignores fallbacks, nothing below can contribute, so stop.
            if (merged != null && merged.ignoresFallbacks()) {
                if (ConfigImpl.traceSubstitutionsEnabled())
                    ConfigImpl.trace(newContext.depth(),
                            "merged ignores fallbacks, skipping remaining stack");
                break;
            }

            // the end value may or may not be resolved already

            ResolveSource sourceForEnd;

            if (end instanceof ReplaceableMergeStack)
                throw new ConfigException.BugOrBroken("A delayed merge should not contain another one: " + replaceable);
            else if (end instanceof Unmergeable) {
                // the remainder could be any kind of value, including another
                // ConfigDelayedMerge
                AbstractConfigValue remainder = replaceable.makeReplacement(context, count + 1);

                if (ConfigImpl.traceSubstitutionsEnabled())
                    ConfigImpl.trace(newContext.depth(), "remainder portion: " + remainder);

                // If, while resolving 'end' we come back to the same
                // merge stack, we only want to look _below_ 'end'
                // in the stack. So we arrange to replace the
                // ConfigDelayedMerge with a value that is only
                // the remainder of the stack below this one.

                if (ConfigImpl.traceSubstitutionsEnabled())
                    ConfigImpl.trace(newContext.depth(), "building sourceForEnd");

                // we resetParents() here because we'll be resolving "end"
                // against a root which does NOT contain "end"
                sourceForEnd = source.replaceWithinCurrentParent((AbstractConfigValue) replaceable, remainder);

                if (ConfigImpl.traceSubstitutionsEnabled())
                    ConfigImpl.trace(newContext.depth(), "  sourceForEnd before reset parents but after replace: "
                            + sourceForEnd);

                sourceForEnd = sourceForEnd.resetParents();
            } else {
                if (ConfigImpl.traceSubstitutionsEnabled())
                    ConfigImpl.trace(newContext.depth(),
                            "will resolve end against the original source with parent pushed");

                sourceForEnd = source.pushParent(replaceable);

                // Same spec rule as the short-circuit above, applied per-key:
                // if every key in the lower-priority 'end' object is already
                // shadowed in 'merged' by a value ignoring fallbacks, the whole
                // 'end' would be discarded by the subsequent merge. Skip
                // resolving it. (We only skip the whole entry — substituting a
                // partial copy of 'end' would change identity and break parent
                // chain walks during inner substitution resolution.)
                if (merged instanceof AbstractConfigObject && end instanceof SimpleConfigObject
                        && allKeysShadowed((SimpleConfigObject) end, (AbstractConfigObject) merged)) {
                    if (ConfigImpl.traceSubstitutionsEnabled())
                        ConfigImpl.trace(newContext.depth(),
                                "all keys in end are shadowed by merged, skipping");
                    count += 1;
                    continue;
                }
            }

            if (ConfigImpl.traceSubstitutionsEnabled()) {
                ConfigImpl.trace(newContext.depth(), "sourceForEnd      =" + sourceForEnd);
            }

            if (ConfigImpl.traceSubstitutionsEnabled())
                ConfigImpl.trace(newContext.depth(), "Resolving highest-priority item in delayed merge " + end
                        + " against " + sourceForEnd + " endWasRemoved=" + (source != sourceForEnd));
            ResolveResult<? extends AbstractConfigValue> result = newContext.resolve(end, sourceForEnd);
            AbstractConfigValue resolvedEnd = result.value;
            newContext = result.context;

            if (resolvedEnd != null) {
                if (merged == null) {
                    merged = resolvedEnd;
                } else {
                    if (ConfigImpl.traceSubstitutionsEnabled())
                        ConfigImpl.trace(newContext.depth() + 1, "merging " + merged + " with fallback " + resolvedEnd);
                    merged = merged.withFallback(resolvedEnd);
                }
            }

            count += 1;

            if (ConfigImpl.traceSubstitutionsEnabled())
                ConfigImpl.trace(newContext.depth(), "stack merged, yielding: " + merged);
        }

        return ResolveResult.make(newContext, merged);
    }

    // True when every key in 'end' is shadowed by a value in 'merged' that
    // ignores fallbacks (so the subsequent merge would drop the whole 'end').
    private static boolean allKeysShadowed(SimpleConfigObject end, AbstractConfigObject merged) {
        if (end.isEmpty())
            return false;
        for (String key : end.keySet()) {
            AbstractConfigValue mergedValue;
            try {
                mergedValue = merged.attemptPeekWithPartialResolve(key);
            } catch (ConfigException.NotResolved e) {
                return false;
            }
            if (mergedValue == null || !mergedValue.ignoresFallbacks())
                return false;
        }
        return true;
    }

    @Override
    public AbstractConfigValue makeReplacement(ResolveContext context, int skipping) {
        return ConfigDelayedMerge.makeReplacement(context, stack, skipping);
    }

    // static method also used by ConfigDelayedMergeObject; end may be null
    static AbstractConfigValue makeReplacement(ResolveContext context, List<AbstractConfigValue> stack, int skipping) {
        List<AbstractConfigValue> subStack = stack.subList(skipping, stack.size());

        if (subStack.isEmpty()) {
            if (ConfigImpl.traceSubstitutionsEnabled())
                ConfigImpl.trace(context.depth(), "Nothing else in the merge stack, replacing with null");
            return null;
        } else {
            // generate a new merge stack from only the remaining items
            AbstractConfigValue merged = null;
            for (AbstractConfigValue v : subStack) {
                if (merged == null)
                    merged = v;
                else
                    merged = merged.withFallback(v);
            }
            return merged;
        }
    }

    @Override
    ResolveStatus resolveStatus() {
        return ResolveStatus.UNRESOLVED;
    }

    @Override
    public AbstractConfigValue replaceChild(AbstractConfigValue child, AbstractConfigValue replacement) {
        List<AbstractConfigValue> newStack = replaceChildInList(stack, child, replacement);
        if (newStack == null)
            return null;
        else
            return new ConfigDelayedMerge(origin(), newStack);
    }

    @Override
    public boolean hasDescendant(AbstractConfigValue descendant) {
        return hasDescendantInList(stack, descendant);
    }

    @Override
    ConfigDelayedMerge relativized(Path prefix) {
        List<AbstractConfigValue> newStack = new ArrayList<AbstractConfigValue>();
        for (AbstractConfigValue o : stack) {
            newStack.add(o.relativized(prefix));
        }
        return new ConfigDelayedMerge(origin(), newStack);
    }

    // static utility shared with ConfigDelayedMergeObject
    static boolean stackIgnoresFallbacks(List<AbstractConfigValue> stack) {
        AbstractConfigValue last = stack.get(stack.size() - 1);
        return last.ignoresFallbacks();
    }

    @Override
    protected boolean ignoresFallbacks() {
        return stackIgnoresFallbacks(stack);
    }

    @Override
    protected AbstractConfigValue newCopy(ConfigOrigin newOrigin) {
        return new ConfigDelayedMerge(newOrigin, stack);
    }

    @Override
    protected final ConfigDelayedMerge mergedWithTheUnmergeable(Unmergeable fallback) {
        return (ConfigDelayedMerge) mergedWithTheUnmergeable(stack, fallback);
    }

    @Override
    protected final ConfigDelayedMerge mergedWithObject(AbstractConfigObject fallback) {
        return (ConfigDelayedMerge) mergedWithObject(stack, fallback);
    }

    @Override
    protected ConfigDelayedMerge mergedWithNonObject(AbstractConfigValue fallback) {
        return (ConfigDelayedMerge) mergedWithNonObject(stack, fallback);
    }

    @Override
    public Collection<AbstractConfigValue> unmergedValues() {
        return stack;
    }

    @Override
    protected boolean canEqual(Object other) {
        return other instanceof ConfigDelayedMerge;
    }

    @Override
    public boolean equals(Object other) {
        // note that "origin" is deliberately NOT part of equality
        if (other instanceof ConfigDelayedMerge) {
            return canEqual(other)
                    && (this.stack == ((ConfigDelayedMerge) other).stack || this.stack
                            .equals(((ConfigDelayedMerge) other).stack));
        } else {
            return false;
        }
    }

    @Override
    public int hashCode() {
        // note that "origin" is deliberately NOT part of equality
        return stack.hashCode();
    }

    @Override
    protected void render(StringBuilder sb, int indent, boolean atRoot, String atKey, ConfigRenderOptions options) {
        render(stack, sb, indent, atRoot, atKey, options);
    }

    @Override
    protected void render(StringBuilder sb, int indent, boolean atRoot, ConfigRenderOptions options) {
        render(sb, indent, atRoot, null, options);
    }

    // static method also used by ConfigDelayedMergeObject.
    static void render(List<AbstractConfigValue> stack, StringBuilder sb, int indent, boolean atRoot, String atKey,
            ConfigRenderOptions options) {
        boolean commentMerge = options.getComments();
        if (commentMerge) {
            sb.append("# unresolved merge of " + stack.size() + " values follows (\n");
            if (atKey == null) {
                indent(sb, indent, options);
                sb.append("# this unresolved merge will not be parseable because it's at the root of the object\n");
                indent(sb, indent, options);
                sb.append("# the HOCON format has no way to list multiple root objects in a single file\n");
            }
        }

        List<AbstractConfigValue> reversed = new ArrayList<AbstractConfigValue>();
        reversed.addAll(stack);
        Collections.reverse(reversed);

        int i = 0;
        for (AbstractConfigValue v : reversed) {
            if (commentMerge) {
                indent(sb, indent, options);
                if (atKey != null) {
                    sb.append("#     unmerged value " + i + " for key "
                            + ConfigImplUtil.renderJsonString(atKey) + " from ");
                } else {
                    sb.append("#     unmerged value " + i + " from ");
                }
                i += 1;
                sb.append(v.origin().description());
                sb.append("\n");

                for (String comment : v.origin().comments()) {
                    indent(sb, indent, options);
                    sb.append("# ");
                    sb.append(comment);
                    sb.append("\n");
                }
            }
            indent(sb, indent, options);

            if (atKey != null) {
                sb.append(ConfigImplUtil.renderJsonString(atKey));
                if (options.getFormatted())
                    sb.append(" : ");
                else
                    sb.append(":");
            }
            v.render(sb, indent, atRoot, options);
            sb.append(",");
            if (options.getFormatted())
                sb.append('\n');
        }
        // chop comma or newline
        sb.setLength(sb.length() - 1);
        if (options.getFormatted()) {
            sb.setLength(sb.length() - 1); // also chop comma
            sb.append("\n"); // put a newline back
        }
        if (commentMerge) {
            indent(sb, indent, options);
            sb.append("# ) end of unresolved merge\n");
        }
    }
}

File: config/src/main/scala/com/typesafe/config/impl/ConfigDelayedMerge.scala
/**
 *   Copyright (C) 2011-2012 Typesafe Inc. <http://typesafe.com>
 */
package com.typesafe.config.impl

import java.{ lang => jl }
import java.{ util => ju }
import scala.collection.JavaConverters._
import com.typesafe.config.ConfigException
import com.typesafe.config.ConfigOrigin
import com.typesafe.config.ConfigRenderOptions
import com.typesafe.config.impl.AbstractConfigValue._

/**
 * The issue here is that we want to first merge our stack of config files, and
 * then we want to evaluate substitutions. But if two substitutions both expand
 * to an object, we might need to merge those two objects. Thus, we can't ever
 * "override" a substitution when we do a merge; instead we have to save the
 * stack of values that should be merged, and resolve the merge when we evaluate
 * substitutions.
 */
object ConfigDelayedMerge {
    // static method also used by ConfigDelayedMergeObject
    @throws[AbstractConfigValue.NotPossibleToResolve]
    def resolveSubstitutions(
        replaceable: ReplaceableMergeStack,
        stack: ju.List[AbstractConfigValue],
        context: ResolveContext,
        source: ResolveSource): ResolveResult[_ <: AbstractConfigValue] = {
        if (ConfigImpl.traceSubstitutionsEnabled) {
            ConfigImpl.trace(
                context.depth,
                "delayed merge stack has " + stack.size + " items:")
            var count = 0
            for (v <- stack.asScala) {
                ConfigImpl.trace(context.depth + 1, count + ": " + v)
                count += 1
            }
        }
        // to resolve substitutions, we need to recursively resolve
        // the stack of stuff to merge, and merge the stack so
        // we won't be a delayed merge anymore. If restrictToChildOrNull
        // is non-null, or resolve options allow partial resolves,
        // we may remain a delayed merge though.
        var newContext = context
        var count = 0
        var merged: AbstractConfigValue = null
        for (end <- stack.asScala) { // the end value may or may not be resolved already
            var sourceForEnd: ResolveSource = null
            if (end.isInstanceOf[ReplaceableMergeStack])
                throw new ConfigException.BugOrBroken(
                    "A delayed merge should not contain another one: " + replaceable)
            else if (end.isInstanceOf[Unmergeable]) { // the remainder could be any kind of value, including another
                // ConfigDelayedMerge
                val remainder =
                    replaceable.makeReplacement(context, count + 1)
                if (ConfigImpl.traceSubstitutionsEnabled)
                    ConfigImpl.trace(
                        newContext.depth,
                        "remainder portion: " + remainder)
                // If, while resolving 'end' we come back to the same
                // merge stack, we only want to look _below_ 'end'
                // in the stack. So we arrange to replace the
                // ConfigDelayedMerge with a value that is only
                // the remainder of the stack below this one.
                if (ConfigImpl.traceSubstitutionsEnabled)
                    ConfigImpl.trace(newContext.depth, "building sourceForEnd")
                // we resetParents() here because we'll be resolving "end"
                // against a root which does NOT contain "end"
                sourceForEnd = source.replaceWithinCurrentParent(
                    replaceable.asInstanceOf[AbstractConfigValue],
                    remainder)
                if (ConfigImpl.traceSubstitutionsEnabled)
                    ConfigImpl.trace(
                        newContext.depth,
                        "  sourceForEnd before reset parents but after replace: " + sourceForEnd)
                sourceForEnd = sourceForEnd.resetParents
            } else {
                if (ConfigImpl.traceSubstitutionsEnabled)
                    ConfigImpl.trace(
                        newContext.depth,
                        "will resolve end against the original source with parent pushed")
                sourceForEnd = source.pushParent(replaceable)
            }
            if (ConfigImpl.traceSubstitutionsEnabled)
                ConfigImpl.trace(newContext.depth, "sourceForEnd      =" + sourceForEnd)
            if (ConfigImpl.traceSubstitutionsEnabled)
                ConfigImpl.trace(
                    newContext.depth,
                    "Resolving highest-priority item in delayed merge " + end
                        + " against " + sourceForEnd + " endWasRemoved=" + (source != sourceForEnd))
            val result =
                newContext.resolve(end, sourceForEnd)
            val resolvedEnd = result.value
            newContext = result.context
            if (resolvedEnd != null)
                if (merged == null) merged = resolvedEnd
                else {
                    if (ConfigImpl.traceSubstitutionsEnabled)
                        ConfigImpl.trace(
                            newContext.depth + 1,
                            "merging " + merged + " with fallback " + resolvedEnd)
                    merged = merged.withFallback(resolvedEnd)
                }
            count += 1
            if (ConfigImpl.traceSubstitutionsEnabled)
                ConfigImpl.trace(newContext.depth, "stack merged, yielding: " + merged)
        }
        ResolveResult.make(newContext, merged)
    }
    // static method also used by ConfigDelayedMergeObject; end may be null
    def makeReplacement(
        context: ResolveContext,
        stack: ju.List[AbstractConfigValue],
        skipping: Int): AbstractConfigValue = {
        val subStack =
            stack.subList(skipping, stack.size)
        if (subStack.isEmpty) {
            if (ConfigImpl.traceSubstitutionsEnabled)
                ConfigImpl.trace(
                    context.depth,
                    "Nothing else in the merge stack, replacing with null")
            null
        } else { // generate a new merge stack from only the remaining items
            var merged: AbstractConfigValue = null
            for (v <- subStack.asScala) {
                if (merged == null) merged = v else merged = merged.withFallback(v)
            }
            merged
        }
    }
    // static utility shared with ConfigDelayedMergeObject
    def stackIgnoresFallbacks(
        stack: ju.List[AbstractConfigValue]): Boolean = {
        val last = stack.get(stack.size - 1)
        last.ignoresFallbacks
    }
    // static method also used by ConfigDelayedMergeObject.
    def render(
        stack: ju.List[AbstractConfigValue],
        sb: jl.StringBuilder,
        indentVal: Int,
        atRoot: Boolean,
        atKey: String,
        options: ConfigRenderOptions): Unit = {
        val commentMerge = options.getComments
        if (commentMerge) {
            sb.append("# unresolved merge of " + stack.size + " values follows (\n")
            if (atKey == null) {
                indent(sb, indentVal, options)
                sb.append(
                    "# this unresolved merge will not be parseable because it's at the root of the object\n")
                indent(sb, indentVal, options)
                sb.append(
                    "# the HOCON format has no way to list multiple root objects in a single file\n")
            }
        }
        val reversed = new ju.ArrayList[AbstractConfigValue]
        reversed.addAll(stack)
        ju.Collections.reverse(reversed)
        var i = 0
        for (v <- reversed.asScala) {
            if (commentMerge) {
                indent(sb, indentVal, options)
                if (atKey != null)
                    sb.append(
                        "#     unmerged value " + i + " for key " + ConfigImplUtil
                            .renderJsonString(atKey) + " from ")
                else sb.append("#     unmerged value " + i + " from ")
                i += 1
                sb.append(v.origin.description)
                sb.append("\n")
                for (comment <- v.origin.comments.asScala) {
                    indent(sb, indentVal, options)
                    sb.append("# ")
                    sb.append(comment)
                    sb.append("\n")
                }
            }
            indent(sb, indentVal, options)
            if (atKey != null) {
                sb.append(ConfigImplUtil.renderJsonString(atKey))
                if (options.getFormatted) sb.append(" : ") else sb.append(":")
            }
            v.render(sb, indentVal, atRoot, options)
            sb.append(",")
            if (options.getFormatted) sb.append('\n')
        }
        // chop comma or newline
        sb.setLength(sb.length - 1)
        if (options.getFormatted) {
            sb.setLength(sb.length - 1) // also chop comma
            sb.append("\n") // put a newline back
        }
        if (commentMerge) {
            indent(sb, indentVal, options)
            sb.append("# ) end of unresolved merge\n")
        }
    }
}

final class ConfigDelayedMerge(
    origin: ConfigOrigin, // earlier items in the stack win
    val stack: ju.List[AbstractConfigValue]) extends AbstractConfigValue(origin)
    with Unmergeable
    with ReplaceableMergeStack {

    if (stack.isEmpty)
        throw new ConfigException.BugOrBroken("creating empty delayed merge value")

    for (v <- stack.asScala) {
        if (v.isInstanceOf[ConfigDelayedMerge] || v
            .isInstanceOf[ConfigDelayedMergeObject])
            throw new ConfigException.BugOrBroken(
                "placed nested DelayedMerge in a ConfigDelayedMerge, should have consolidated stack")
    }

    override def valueType = throw new ConfigException.NotResolved(
        "called valueType() on value with unresolved substitutions, need to Config#resolve() first, see API docs")

    override def unwrapped = throw new ConfigException.NotResolved(
        "called unwrapped() on value with unresolved substitutions, need to Config#resolve() first, see API docs")

    @throws[AbstractConfigValue.NotPossibleToResolve]
    override def resolveSubstitutions(
        context: ResolveContext,
        source: ResolveSource): ResolveResult[_ <: AbstractConfigValue] =
        ConfigDelayedMerge.resolveSubstitutions(this, stack, context, source)

    override def makeReplacement(
        context: ResolveContext,
        skipping: Int): AbstractConfigValue =
        ConfigDelayedMerge.makeReplacement(context, stack, skipping)

    override def resolveStatus: ResolveStatus = ResolveStatus.UNRESOLVED

    override def replaceChild(
        child: AbstractConfigValue,
        replacement: AbstractConfigValue): AbstractConfigValue = {
        val newStack =
            AbstractConfigValue.replaceChildInList(stack, child, replacement)
        if (newStack == null) null else new ConfigDelayedMerge(origin, newStack)
    }
    override def hasDescendant(descendant: AbstractConfigValue): Boolean =
        AbstractConfigValue.hasDescendantInList(stack, descendant)

    override def relativized(prefix: Path): ConfigDelayedMerge = {
        val newStack = new ju.ArrayList[AbstractConfigValue]
        for (o <- stack.asScala) {
            newStack.add(o.relativized(prefix))
        }
        new ConfigDelayedMerge(origin, newStack)
    }

    override def ignoresFallbacks: Boolean =
        ConfigDelayedMerge.stackIgnoresFallbacks(stack)

    override def newCopy(newOrigin: ConfigOrigin) =
        new ConfigDelayedMerge(newOrigin, stack)

    override final def mergedWithTheUnmergeable(
        fallback: Unmergeable): ConfigDelayedMerge =
        mergedWithTheUnmergeable(stack, fallback)
            .asInstanceOf[ConfigDelayedMerge]

    override final def mergedWithObject(
        fallback: AbstractConfigObject): ConfigDelayedMerge =
        mergedWithObject(stack, fallback).asInstanceOf[ConfigDelayedMerge]

    override def mergedWithNonObject(
        fallback: AbstractConfigValue): ConfigDelayedMerge =
        mergedWithNonObject(stack, fallback).asInstanceOf[ConfigDelayedMerge]

    override def unmergedValues: ju.Collection[AbstractConfigValue] = stack

    override def canEqual(other: Any): Boolean =
        other.isInstanceOf[ConfigDelayedMerge]

    override def equals(other: Any): Boolean = {
        // note that "origin" is deliberately NOT part of equality
        if (other.isInstanceOf[ConfigDelayedMerge])
            canEqual(other) && ((this.stack eq other
                .asInstanceOf[ConfigDelayedMerge]
                .stack) || this.stack == other.asInstanceOf[ConfigDelayedMerge].stack)
        else false
    }

    override def hashCode: Int = stack.hashCode

    override def render(
        sb: jl.StringBuilder,
        indent: Int,
        atRoot: Boolean,
        atKey: String,
        options: ConfigRenderOptions): Unit = {
        ConfigDelayedMerge.render(stack, sb, indent, atRoot, atKey, options)
    }

    override def render(
        sb: jl.StringBuilder,
        indent: Int,
        atRoot: Boolean,
        options: ConfigRenderOptions): Unit = {
        render(sb, indent, atRoot, null, options)
    }
}
